spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraCatalog.scala

```diff
@@ -1,65 +1,157 @@
-    val clusterOpt = toOption(cluster)
-    val tableWithQualifiers = Subquery(table, CassandraRelation(tableDef, alias, clusterOpt)(cc))
-    alias.map(a => Subquery(a, tableWithQualifiers)).getOrElse(tableWithQualifiers)
+    val tableIdent = fullTableIdent(tableIdentifier)
+    val tableLogicPlan = cachedDataSourceTables.get(tableIdent)
+    alias.map(a => Subquery(a, tableLogicPlan)).getOrElse(tableLogicPlan)
   }
 
-  private def toOption(cluster: String): Option[String] = {
-    cluster match {
-      case "default" => None
-      case _         => Option(cluster)
-    }
+  /** Build logic plan from a CassandraSourceRelation */
+  private def buildRelation(tableIdentifier: Seq[String]): LogicalPlan = {
+    val (cluster, database, table) = getClusterDBTableNames(tableIdentifier)
+    val tableRef = TableRef(table, database, Option(cluster))
+    val sourceRelation = CassandraSourceRelation(tableRef, csc, CassandraSourceOptions())
+    Subquery(table, LogicalRelation(sourceRelation))
   }
 
+  /** Return cluster, database and table names from a table identifier*/
   private def getClusterDBTableNames(tableIdentifier: Seq[String]): (String, String, String) = {
     val id = processTableIdentifier(tableIdentifier).reverse.lift
-    (id(2).getOrElse("default"), id(1).getOrElse(cc.getKeyspace), id(0).getOrElse(throw new IOException(s"Missing table name")))
+    val cluster = id(2).getOrElse(csc.getCluster)
+    val database = id(1).getOrElse(csc.getKeyspace)
+    val table = id(0).getOrElse(throw new IOException(s"Missing table name"))
+    (cluster, database, table)
+  }
+
+  /** Return a table identifier with table name, keyspace name and cluster name */
+  private def fullTableIdent(tableIdentifier: Seq[String]) : Seq[String] = {
+    val (cluster, database, table) = getClusterDBTableNames(tableIdentifier)
+    Seq(cluster, database, table)
   }
 
-  override def registerTable(tableIdentifier: Seq[String], plan: LogicalPlan): Unit = ???
+  override def registerTable(tableIdentifier: Seq[String], plan: LogicalPlan): Unit = {
+    val tableIdent = fullTableIdent(tableIdentifier)
+    cachedDataSourceTables.put(tableIdent, plan)
+  }
 
-  override def unregisterTable(tableIdentifier: Seq[String]): Unit = ???
+  override def unregisterTable(tableIdentifier: Seq[String]): Unit = {
+    val tableIdent = fullTableIdent(tableIdentifier)
+    cachedDataSourceTables.invalidate(tableIdent)
+  }
 
-  override def unregisterAllTables(): Unit = ???
+  override def unregisterAllTables(): Unit = {
+    cachedDataSourceTables.invalidateAll()
+  }
 
   override def tableExists(tableIdentifier: Seq[String]): Boolean = {
     val (cluster, database, table) = getClusterDBTableNames(tableIdentifier)
-    val schema = schemas.get(cluster)
-    val tabDef = for (ksDef <- schema.keyspaceByName.get(database); tabDef <- ksDef.tableByName.get(table)) yield tabDef
-    tabDef.isDefined
+    val tableRef = TableRef(table, database, Option(cluster))
```

Comment:
tableRef is used only on one branch of the condition below, you can move it down